wallet/wallet-cli/src/commands/mod.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use clap::Parser;
+use node_comm::node_traits::{NodeInterface, PeerId};
+use reedline::Reedline;
+use serialization::hex::HexEncode;
+use wallet::DefaultWallet;
+
+use crate::{cli_println, console::ConsoleContext, errors::WalletCliError};
+
+#[derive(Debug, Parser)]
+#[clap(rename_all = "lower")]
+pub enum WalletCommands {
+    /// Returns the current best block hash
+    BestBlock,
+
+    /// Returns the current best block height
+    BestBlockHeight,
+
+    /// Submit a block to be included in the chain
+    ///
+    /// More information about block submits.
+    /// More information about block submits.
+    ///
+    /// Even more information about block submits.
+    /// Even more information about block submits.
+    /// Even more information about block submits.
+    /// Even more information about block submits.
+    SubmitBlock {
+        /// Hex encoded block
+        block: String,
+    },
+
+    SubmitTransaction {
+        /// Hex encoded transaction
+        transaction: String,
+    },
+
+    /// Rescan
+    Rescan,
+
+    /// Node version
+    NodeVersion,
+
+    /// Node shutdown
+    NodeShutdown,
+
+    /// Connect to the remote peer
+    Connect { address: String },
+
+    /// Disconnected the remote peer
+    Disconnect { peer_id: PeerId },
+
+    /// Get connected peer count
+    PeerCount,
+
+    /// Get connected peers
+    ConnectedPeers,
+
+    /// Add reserved peer
+    AddReservedPeer { address: String },
+
+    /// Remove reserved peer
+    RemoveReservedPeer { address: String },
+
+    /// Quit the REPL
+    Exit,
+
+    /// Print history
+    History,
+
+    /// Clear screen
+    #[clap(name = "clear")]
+    ClearScreen,
+
+    /// Clear history
+    ClearHistory,
+}
+
+pub async fn handle_wallet_command(
```

Comment:
This function has to be part of the controller. Not in the sense that it has to be moved, but in the sense that the controller has to expose function calls (maybe through a trait?) with all the functionality of the wallet that a "client" (whether GUI or not) should be able to do. Then, this function will parse that command here in this crate, in this function, and call that function in the controller to do the task.

In other words: This crate should not even know that there's something called "NodeInterface". So node-comm should not be here.

Reply:
Btw, this is called "model-view-controller" design.

So that trait would be the model, and this crate would be the view. Hopefully I'm not missing something :smile:

Reply:
I added a simple `Controller` for this. For now it's just a wrapper for `NodeInterface`, but it will be extended later to do more sophisticated stuff.